tests/UnifiedSpecTests/Context.php

```diff
@@ -0,0 +1,395 @@
+            assertInstanceOf(Session::class, $session);
+            $args['session'] = $session;
+        }
+
+        if (array_key_exists('writeConcern', $args)) {
+            assertInternalType('object', $args['writeConcern']);
+            $args['writeConcern'] = self::prepareWriteConcern($args['writeConcern']);
+        }
+
+        return $args;
+    }
+
+    public function startEventObservers()
+    {
+        foreach ($this->eventObserversByClient as $eventObserver) {
+            $eventObserver->start();
+        }
+    }
+
+    public function stopEventObservers()
+    {
+        foreach ($this->eventObserversByClient as $eventObserver) {
+            $eventObserver->stop();
+        }
+    }
+
+    private static function assertHasOnlyKeys($arrayOrObject, array $keys)
+    {
+        assertThat($arrayOrObject, logicalOr(isType('array'), isType('object')));
+        $diff = array_diff_key((array) $arrayOrObject, array_fill_keys($keys, 1));
+        assertEmpty($diff, 'Unsupported keys: ' . implode(',', array_keys($diff)));
+    }
+
+    private function createClient(stdClass $o) : Client
+    {
+        self::assertHasOnlyKeys($o, ['id', 'uriOptions', 'useMultipleMongoses', 'observeEvents', 'ignoreCommandMonitoringEvents']);
+
+        $useMultipleMongoses = $o->useMultipleMongoses ?? null;
+        $observeEvents = $o->observeEvents ?? null;
+        $ignoreCommandMonitoringEvents = $o->ignoreCommandMonitoringEvents ?? [];
+
+        $uri = $this->uri;
+
+        if (isset($useMultipleMongoses)) {
+            assertInternalType('bool', $useMultipleMongoses);
+
+            if ($useMultipleMongoses) {
+                self::requireMultipleMongoses($uri);
+            } else {
+                $uri = self::removeMultipleMongoses($uri);
+            }
+        }
+
+        $uriOptions = [];
+
+        if (isset($o->uriOptions)) {
+            assertInternalType('object', $o->uriOptions);
+            /* TODO: If readPreferenceTags is set, assert it is an array of
+             * strings and convert to an array of documents expected by the
+             * PHP driver. */
+            $uriOptions = (array) $o->uriOptions;
+        }
+
+        if (isset($observeEvents)) {
+            assertInternalType('array', $observeEvents);
+            assertInternalType('array', $ignoreCommandMonitoringEvents);
+
+            $this->eventObserversByClient[$o->id] = new EventObserver($observeEvents, $ignoreCommandMonitoringEvents);
+        }
+
+        return new Client($uri, $uriOptions);
+    }
+
+    private function createCollection(stdClass $o) : Collection
+    {
+        self::assertHasOnlyKeys($o, ['id', 'database', 'collectionName', 'collectionOptions']);
+
+        $collectionName = $o->collectionName ?? null;
+        $database = $o->database ?? null;
+
+        assertInternalType('string', $collectionName);
+        assertInternalType('string', $database);
+        assertArrayHasKey($database, $this->entityMap);
```

Comment:
Just wanted to point out that the entity map does this check already, but also provides a more helpful error message ("No entity is defined for ...") as opposed to a simple ("Failed asserting that an array has the key ...").

I think we can remove the check on the entity map here (and in all occurrences that follow) and defer to the map itself to perform the right checks.

Reply:
Also removing the `assertArrayNotHasKey` in `createEntities`, since EntityMap enforces that we cannot overwrite entities.